### PR TITLE
T/149: Release task can't find a changelog so it can't publish a release

### LINKED ISSUE #99: Release task fails if there's no CHANGELOG.md
This may happen when releasing some package for the first time. We either need to allow releasing packages if one of them doesn't have a CHANGELOG or a way to generate changelog for that package only. The latter is problematic because we need to publish the package on npm before we start generating changelog for other packages because without any version published on npm that package cannot be added to ckeditor5 deps (CI fails to install it).

### LINKED ISSUE #147: If there is not tag in the repo release:dependencies throws
```
(master c80166f) p@m /workspace/ckeditor5> gulp release:dependencies
[18:37:35] Using gulpfile /workspace/ckeditor5/gulpfile.js
[18:37:35] Starting 'release:dependencies'...
[18:37:36] Error while executing git describe --abbrev=0 --tags 2> /dev/null:
```

### LINKED ISSUE #149: Release task can't find a changelog so it can't publish a release
```
[18:51:36] Using gulpfile /workspace/ckeditor5/gulpfile.js
[18:51:36] Starting 'release:dependencies'...
? Packages to release:
  * "@ckeditor/ckeditor5-autoformat": 0.4.1 => v0.5.0
  * "@ckeditor/ckeditor5-basic-styles": 0.7.1 => v0.8.0
  * "@ckeditor/ckeditor5-block-quote": 0.0.0 => v0.1.0
  * "@ckeditor/ckeditor5-build-classic": 0.0.1 => v0.1.0
  * "@ckeditor/ckeditor5-clipboard": 0.4.1 => v0.5.0
  * "@ckeditor/ckeditor5-core": 0.7.0 => v0.8.0
  * "@ckeditor/ckeditor5-editor-classic": 0.7.1 => v0.7.2
  * "@ckeditor/ckeditor5-editor-inline": 0.0.1 => v0.1.0
  * "@ckeditor/ckeditor5-engine": 0.8.0 => v0.9.0
  * "@ckeditor/ckeditor5-enter": 0.8.0 => v0.9.0
  * "@ckeditor/ckeditor5-heading": 0.8.0 => v0.9.0
  * "@ckeditor/ckeditor5-image": 0.4.0 => v0.5.0
  * "@ckeditor/ckeditor5-link": 0.5.1 => v0.6.0
  * "@ckeditor/ckeditor5-list": 0.5.1 => v0.6.0
  * "@ckeditor/ckeditor5-markdown-gfm": 0.4.1 => v0.4.2
  * "@ckeditor/ckeditor5-paragraph": 0.6.1 => v0.7.0
  * "@ckeditor/ckeditor5-presets": 0.1.1 => v0.2.0
  * "@ckeditor/ckeditor5-theme-lark": 0.6.1 => v0.7.0
  * "@ckeditor/ckeditor5-typing": 0.8.0 => v0.9.0
  * "@ckeditor/ckeditor5-ui": 0.7.1 => v0.8.0
  * "@ckeditor/ckeditor5-undo": 0.7.1 => v0.8.0
  * "@ckeditor/ckeditor5-utils": 0.8.0 => v0.9.0
  * "@ckeditor/ckeditor5-widget": 0.0.1 => v0.1.0
Continue? Yes
? Select services where packages will be released: (Press <space> to select, <a> to toggle all, <i> to inverse selection)npm, GitHub
? Provide the GitHub token: ****************************************
[18:53:08] Creating release for "@ckeditor/ckeditor5-autoformat".
[18:53:08] Updating dependencies...
[18:53:11] Committing "package.json"...
[18:53:13] Creating a tag...
[18:53:17] Publishing on NPM...
[18:53:22] Creating a GitHub release...
[18:53:23] Release "v0.5.0" has been created and published.

[18:53:23] Creating release for "@ckeditor/ckeditor5-basic-styles".
[18:53:24] Updating dependencies...
[18:53:26] Committing "package.json"...
[18:53:28] Creating a tag...
[18:53:32] Publishing on NPM...
[18:53:35] Creating a GitHub release...
[18:53:36] Release "v0.8.0" has been created and published.

[18:53:36] Creating release for "@ckeditor/ckeditor5-block-quote".
[18:53:36] Updating dependencies...
[18:53:39] Cannot find changelog entries for 0.1.0.
[18:53:39] Finished 'release:dependencies' after 2.03 min
(master c80166f) p@m /workspace/ckeditor5>
```

Most likely related to https://github.com/ckeditor/ckeditor5-dev/issues/148

But the real bug here is that this threw **while** releasing ;/ The task must not look for anything or do anything that may break during releasing.

### PR DESCRIPTION
**Pull request by pomek**
### Suggested merge commit message ([convention](https://github.com/ckeditor/ckeditor5-design/wiki/Git-commit-message-convention))

Fix: Fixed issues related to broken release process when any error occurs. Closes #99. Closes #147.  Closes: #149. Closes: #151.

Release tool now does not read required parameters itself. You have to pass the parameters (like version, description of the changes, etc) manually.

---

### Additional information

~Most probably the #151 will be also resolved because if any error occurs during release a package, the release process will not be terminated. I also added additional validation - if the changelog entries were not generated before the releasing - release won't start.~

### CODE REVIEW

**Comment by Reinmar** [comment]
Mmm... I'm not sure what the patch does really. And whether it fixes the problem for real. The process should never fail due to predictable reasons. All the validation should be done beforehand, all the data read. Then the process should throw. Loud and fast. But during releasing the script may only fail if e.g. internet connection was lost. But in such case, I'd actually like the process to fail ASAP and terminate. Why? Because I'll have less mess to clean up.



> Most probably the #151 will be also resolved because if any error occurs during release a package, the release process will not be terminated. I also added additional validation - if the changelog entries were not generated before the releasing - release won't start.

This wasn't the issue that I reported. The issue I reported is that all dependencies must be updated to their latest versions – not only these which are being released.

**Comment by pomek** [comment]
I've just updated the suggested merge commit message.

@Reinmar, you can review it once again. Release tool works now as you described above.

**Comment by Reinmar** [comment]
OK, I finally had a moment to look into this.

In general, the changes are quite straightforward and make sense. They make the process more predictable and safe. However, few things still need polishing.

* [ ] lib/release-tools/tasks/releaserepository.js
  * `options.changes` and `options.version` were added but how are we supposed to retrieve this information? And in what format? How one can use this task when it can't easily retrieve these values?
  * `options` are not validated in any way – the validation needs to be done outside. But taken that `releasesubrepositories.js` makes it internally... how would you reuse this code? 
* [ ] lib/release-tools/tasks/releasesubrepositories.js
  * The biggest problem with this code is that:
    * there are three global state variables which are modified in many points of the process so you have no idea when and how they get modified – `packagesToRelease`, `dependencies` and `errors`,
    * most of these functions do different things than what their names say – `displayAndConfirmReleasedPackages()` modifies `packagesToRelease`, `prepareDepdenciesVersion()` handles lack of user's confirmation, `getReleaseOptions()` calls `configureReleaseOptions()` (the names don't match because the latter is incorrect) but also throws an unrelated error, etc. etc.
    * naming is unclear – e.g. `updateDependenciesForReleasedPackages()` – those packages are not yet released and it's not "for" but "of".
* [x] lib/release-tools/utils/changelog.js
* [x] lib/release-tools/{tasks => utils}/creategithubrelease.js
* [x] lib/release-tools/utils/getpackagejson.js
* [x] lib/release-tools/utils/getpackagestorelease.js
* [x] lib/release-tools/utils/releasevalidator.js
* [x] lib/release-tools/utils/updatedependenciesversions.js
* [x] lib/release-tools/utils/versions.js
* [ ] tests/release-tools/tasks/generatechangelogforsinglepackage.js
* [ ] tests/release-tools/tasks/generatechangelogforsubpackages.js
* [ ] tests/release-tools/tasks/generatechangelogforsubrepositories.js
* [ ] tests/release-tools/tasks/releaserepository.js
* [ ] tests/release-tools/tasks/releasesubrepositories.js
* [ ] tests/release-tools/tasks/stubs/releasesubrepositories/package.json
* [ ] tests/release-tools/tasks/stubs/releasesubrepositories/packages/alpha/package.json
* [ ] tests/release-tools/tasks/stubs/releasesubrepositories/packages/beta/package.json
* [ ] tests/release-tools/tasks/stubs/releasesubrepositories/packages/delta/package.json
* [ ] tests/release-tools/tasks/stubs/releasesubrepositories/packages/epsilon/package.json
* [ ] tests/release-tools/tasks/stubs/releasesubrepositories/packages/gamma/package.json
* [ ] tests/release-tools/utils/changelog.js
* [ ] tests/release-tools/{tasks => utils}/creategithubrelease.js
* [ ] tests/release-tools/utils/generatechangelogfromcommits.js
* [ ] tests/release-tools/utils/getnewreleasetype.js
* [ ] tests/release-tools/utils/getpackagestorelease.js
* [ ] tests/release-tools/utils/getsubpackagespaths.js
* [ ] tests/release-tools/utils/getsubrepositoriespaths.js
* [ ] tests/release-tools/utils/releasevalidator.js
* [ ] tests/release-tools/utils/updatedependenciesversions.js

**Comment by pomek** [comment]
> * [ ] lib/release-tools/tasks/releaserepository.js
>  * `options.changes` and `options.version` were added but how are we supposed to retrieve this information? And in what format? How one can use this task when it can't easily retrieve these values?
>  * `options` are not validated in any way – the validation needs to be done outside. But taken that `releasesubrepositories.js` makes it internally... how would you reuse this code? 

Done. Required options will be read by the function itself. Also, the options will be validated by the function.

> * [ ] lib/release-tools/tasks/releasesubrepositories.js
>   * The biggest problem with this code is that:
>     * there are three global state variables which are modified in many points of the process so you have no idea when and how they get modified – `packagesToRelease`, `dependencies` and `errors`,
>     * most of these functions do different things than what their names say – `displayAndConfirmReleasedPackages()` modifies `packagesToRelease`, `prepareDepdenciesVersion()` handles lack of user's confirmation, `getReleaseOptions()` calls `configureReleaseOptions()` (the names don't match because the latter is incorrect) but also throws an unrelated error, etc. etc.
>     * naming is unclear – e.g. `updateDependenciesForReleasedPackages()` – those packages are not yet released and it's not "for" but "of".

I tried to simplify the code, but the _global_ state variables must leave as they are. Some required data are built during the code execution. In order to avoid the _global_ variables, we can pass it directly to the `.then()` callbacks but IMO it's overengineered and will be harder to maintain.

Ready to review once again.

File: packages/ckeditor5-dev-env/lib/release-tools/utils/cli.js [hunk 1]
```diff
@@ -34,7 +34,7 @@ const cli = {
 		};
 
 		return inquirer.prompt( [ confirmQuestion ] )
-			.then( answers => answers.confirm );
+			.then( answers => answers.confirm ? Promise.resolve() : Promise.reject() );
```
Comment (Reinmar) [inline packages/ckeditor5-dev-env/lib/release-tools/utils/cli.js]: This is like throwing an error. Error means that something didn't work. So this change is incorrect.

Comment (Reinmar) [inline packages/ckeditor5-dev-env/lib/release-tools/utils/cli.js]: I'll merge this PR anyway (if I won't find other issues). But please fix it anyway.

**Comment by Reinmar** [comment]
There's still one thing which crashes in the middle of the process – and this is generating changelog for these packages which didn't have them. But that's a lot harder to avoid.

Other than that, I ran the code on CKEditor 5 by disabling a few things and I think that it did what it was supposed to so I hope it's all ok.

**Comment by pomek** [comment]
> There's still one thing which crashes in the middle of the process – and this is generating changelog for these packages which didn't have them. But that's a lot harder to avoid.

Could you prepare a step list which allows me reproduced the issue? I'm not sure whether I understand correctly.

**Comment by Reinmar** [comment]
1. Make sure that for some reason committing a changelog to one of the packages that didn't have any changes since the last release fails.
2. Try to release CKEditor 5.
3. The release process will crash.

Of course, if committing or pushing to repos crashes then the whole release process is endangered. So that's why I wrote that there's nothing we can do with this. We need to generate those empty changelog entries for packages which had only updated dependencies at this stage and this may fail. Fortunately, the chance is very low. So, don't worry. It was just an observation.

**Pull request merged**